### DISCUSSION TITLE
Decoding ABI Error entities

### DISCUSSION THREAD

**Discussion by dalechyn** [wevm/wagmi#1969] [Ideas]
Since we can declare custom errors in Solidity, there probably should be a way to decode them like `useContractError`?

Considering that we already have such data extracted in the generated by `@wagmi/cli` abis:

```
  {
    type: 'error',
    inputs: [],
    name: 'CustomError1'
  },
  {
    type: 'error',
    inputs: [
      { name: 'amount', internalType: 'uint256', type: 'uint256' }
    ],
    name: 'CustomError2'
  },
```
it would be super cool to have a hook to decode them.

Example usage could be the next:
```ts
const { write, error } = useContractWrite(...)

const { data } = useContractError(error)
```

Or there could be a `CustomError` class that would in fact dynamically build `this.name` out of the defined ABI Error's `name` parameter, along with parsed arguments.

```ts
const { write, error } = useContractWrite(...)

console(error.message) // {call,gas estimation} failed with `CustomError2` error with next parameters: `amount: 1337`
```

**Comment by jxom** [comment]
Expect to see this in wagmi v1. 

Our new library [viem.sh](https://viem.sh) has some smarts to [handle custom Solidity errors](https://github.com/wagmi-dev/viem/blob/main/src/errors/contract.test.ts#L365-L380). :)

Idea is that you'll be able to retrieve contract data/errors from write functions using something in wagmi v1 that wraps [`simulateContract`](https://www.viem.sh/docs/contract/simulateContract.html) – where custom errors are automatically unwrapped.

**Reply by jxom** [reply]
If you want to use this now, you can import `ContractFunctionRevertedError` from `viem`, and [insert your ABI, error data and the contract function you called](https://github.com/wagmi-dev/viem/blob/main/src/errors/contract.test.ts#L365-L380) ;)

You can also use the more low-level [`decodeErrorResult`](https://www.viem.sh/docs/contract/decodeErrorResult.html)!

**Reply by dalechyn** [reply]
Oh! Gotcha thanks!
